File: leetcode/0001-1000/0201-0300/0236-lowest-common-ancestor-of-a-binary-tree/LowestCommonAncestorOfABinaryTree.java
public class LowestCommonAncestorOfABinaryTree {
  public static void main(String[] args) {
    //
  }
}

class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;

  TreeNode(int x) {
    val = x;
  }
}

class Solution {
  public TreeNode lowestCommonAncestor(TreeNode node, TreeNode p, TreeNode q) {
    if (node == null || node == p || node == q) {
      return node;
    }

    TreeNode left = lowestCommonAncestor(node.left, p, q);
    TreeNode right = lowestCommonAncestor(node.right, p, q);

    if (left == null) {
      return right;
    }

    if (right == null) {
      return left;
    }

    return node;
  }
}

File: leetcode/0001-1000/0201-0300/0231-0240/0236-lowest-common-ancestor-of-a-binary-tree/LowestCommonAncestorOfABinaryTree.ts
class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
  }
}

function lowestCommonAncestor(node: TreeNode | null, p: TreeNode | null, q: TreeNode | null): TreeNode | null {
  if (node == null || node == p || q == node) {
    return node;
  }

  const left = lowestCommonAncestor(node.left, p, q);
  const right = lowestCommonAncestor(node.right, p, q);

  if (left == null) {
    return right;
  } else if (right == null) {
    return left;
  }

  return node;
};
